### DISCUSSION TITLE
Setting Starlette's max part size for form data

### DISCUSSION THREAD

**Discussion by olliestanley** [fastapi/fastapi#12961] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from fastapi import FastAPI
app = FastAPI()

@app.put("/upload")
async def upload_document(content: Annotated[str, Form()]):
    # Do anything with `content`
    pass
```

### Description

- Since Starlette 0.40.0 their `MultiPartParser` has a new attribute `max_part_size` to resolve a security vulnerability
- This defaults to 1024KB
- This seems to be the cause of an error (HTTP 400) when a form parameter exceeding this size is passed
- The error message is `Part exceeded maximum size of 1024KB.`
- I have verified this error does not occur using older Starlette/FastAPI versions

Of course, this is intended behaviour. What I would like to know is how we can change the value of `max_part_size` in the `MultiPartParser` being used under the hood when using FastAPI with form parameters rather than when using Starlette directly?

Thanks!

### Operating System

Linux

### FastAPI Version

0.115.5

### Pydantic Version

2.9.1

### Python Version

3.10

**Comment by estebanx64** [comment]
Hi there! 👋

Is there any special reason to use a form data with more than 1024kb?

Seeing your endpoint definition as `upload_document` we may be using the wrong approach, if you are trying to upload a document the `File` and `UploadFile` are the solutions for it. 🤓

https://fastapi.tiangolo.com/tutorial/request-files/#import-file

Otherwise, can you share with us more details? So we can then seek a better solution 🚀

**Reply by olliestanley** [reply]
> Seeing your endpoint definition as `upload_document` we may be using the wrong approach, if you are trying to upload a document the `File` and `UploadFile` are the solutions for it. 🤓

Fair point.

The actual endpoint in our server does primarily support file upload, but we added an alternative raw text field for API callers who wished to upload content without using a file. This is often when they have plain text extracted from a webpage, for example. Some of our users have now started coming across the 1024KB limit when using particularly lengthy webpages.

If the answer to this is just that there isn't a supported way to modify the `max_part_size` when working with FastAPI, that's fine. We can require that our users upload a file instead. If there is a simple way to increase it to something like 4096KB though, that would be great.

**Reply by luzzodev** [reply]
> > Seeing your endpoint definition as `upload_document` we may be using the wrong approach, if you are trying to upload a document the `File` and `UploadFile` are the solutions for it. 🤓
> 
> Fair point.
> 
> The actual endpoint in our server does primarily support file upload, but we added an alternative raw text field for API callers who wished to upload content without using a file. This is often when they have plain text extracted from a webpage, for example. Some of our users have now started coming across the 1024KB limit when using particularly lengthy webpages.
> 
> If the answer to this is just that there isn't a supported way to modify the `max_part_size` when working with FastAPI, that's fine. We can require that our users upload a file instead. If there is a simple way to increase it to something like 4096KB though, that would be great.

There are for sure a better way to pass file content then in a form field.

By the way there are no params to change this behaviour but ungly way bypass is to override the class value globally before declaring your fastapi app.

```
from starlette.formparsers import MultiPartParser

MultiPartParser.max_part_size = 1024 * 1024 # 1Kb
```

**Reply by estebanx64** [reply]
The solution given by @luzzodev can be a workaround for it.

You might need to modify max_file_size as well depending on your needs.

```python
from starlette.formparsers import MultiPartParser

MultiPartParser.max_part_size = 10 * 1024 * 1024  # 10MB
MultiPartParser.max_file_size = 20 * 1024 * 1024   # 20MB
```

**Reply by wordhui** [reply]
> The solution given by @luzzodev can be a workaround for it.
> 
> You might need to modify max_file_size as well depending on your needs.
> 
> ```python
> from starlette.formparsers import MultiPartParser
> 
> MultiPartParser.max_part_size = 10 * 1024 * 1024  # 10MB
> MultiPartParser.max_file_size = 20 * 1024 * 1024   # 20MB
> ```

The value of `MultiPartParser.max_file_size` should be used for `SpooledTemporaryFile(max_size=self.max_file_size)`, it seems that you don't need to set this parameter, if you set it to 20M, then anything up to 20M will be saved in memory.

**Reply by reith** [reply]
This has stopped working since Starlette 0.44.0 with the change introduced in [this commit](https://github.com/encode/starlette/pull/2815).

**Reply by armant** [reply]
Is there a new workaround for this now that this has stopped working since Starlette 0.44.0?

**Reply by luzzodev** [reply, marked as answer]
Hello @armant,

Since Starlette has introduced parameters to control form parsing options: `max_files`, `max_fields`, and `max_part_size`. It's now possible to intercept the request with a middleware and call request.form() with the desired options.
Parsed form data is computed only once (and cached), so you won’t encounter any issues later when accessing the form through FastAPI.
The advantage of this approach is that you can control the size of each form dynamically, based on the incoming request.

That said, there's no doubt we should have a way to control these parameters globally.

**Reply by Nixxs** [reply]
Hi @luzzodev  I'm running into this issue right now and I'm sure I don't understand this but when I add this middleware:

```python
from fastapi import Request
from starlette.middleware.base import BaseHTTPMiddleware

class PayloadSizeIncrease(BaseHTTPMiddleware):
    async def dispatch(self, request: Request, call_next):
        # do something with the request object, for example
        max_mb = 50
        max_bytes = max_mb * 1024 * 1024
        form = await request.form(
            max_part_size=max_bytes
		)
        
        # process the request and get the response    
        response = await call_next(request)
        
        return response
```

I lose all the form data, I don't understand this flow very well hoping you can shed some light on how to do this properly?

**Reply by luzzodev** [reply]
Hello @Nixxs, you are using BaseHttMiddleware and this is the reason why you are losing you form data. You can find more information here https://github.com/fastapi/fastapi/discussions/13739.
Hope this helps 😁

**Reply by alex-700** [reply]
Hi, @luzzodev !

> It's now possible to intercept the request with a middleware and call request.form() with the desired options.

Could you elaborate more here? e.g. could you provide full example of the workaround based on your idea of calling `Request.form` inside a middleware?

---

Thoughts:
- We could not do this via `starlette` usual middleware (ASGI app interface), because there is no "shared" `Request` there, and if we create one, and call `form()` that we just drain the stream, and underlying apps recieve nothing.
- Using `@app.middleware('http')` is a bit discouraging for this too. The `Request` here is "shared", but not always, so again calling `form()` stores the form data in `Request._form`, but drains the stream, and so newly created `Request` inside api router gets nothing from the stream.

**Reply by luzzodev** [reply]
Hello @alex-700, I’m actually on mobile, so I’m unable to provide a full example, but I’ll try to give you some ideas.

I totally agree with you and with what you have pointed out.

Starting from the most important one:

> That said, there’s no doubt we should have a way to control these parameters globally.

My opinion on this is that FastAPI must provide an easy way to control all limits and constraints related to forms at all levels (globally, per route, etc.).

Before the changes, you were able to control the limits globally with the first suggested workaround. After the changes, you now have a way to control those constraints without any workaround. On the other hand, FastAPI still does not offer a built-in way to control this, and you have to use different approaches depending on the case.

1 You can refer to https://fastapi.tiangolo.com/how-to/custom-request-and-route/#create-a-custom-gziprequest-class

And create a custom request with the proper handling of the form. 


2 You can use only the Request dependency and access to form with custom constrains ( this one will not work in some cases and you should build the doc manually 

3 you can create a custom middleware taking inspiration from the base http middleware and forward the stream or cache  it.

All those approaches have their own pitfalls and there is not a way that will fit all use cases, especially if you need to have the the form constrains more relaxed.

As far as I know, there isn’t a two-line solution for that, and like I said at the beginning, we should have a way to control it.

Hope this helps

**Reply by pandeyshekhar** [reply]
Hello @luzzodev Please provide a complete working example. I am unable to set a file upload limit and handle the exception when it is exceeded.

**Reply by JaeWangL** [reply]
@pandeyshekhar

### helper
```python
from starlette.datastructures import FormData
from starlette.formparsers import MultiPartParser
from starlette.responses import FileResponse
from starlette.requests import Request

async def form_body_with_large_limits(request: Request) -> FormData:
    try:
        parser = MultiPartParser(
            headers=request.headers,
            stream=request.stream(),
            max_part_size=50 * 1024 * 1024, # 50MB
        )
        form_data = await parser.parse()
        return form_data
    except Exception as e:
        raise e
 ```
 
 ### example
 ```python
 async def upload_zip(
    background_tasks: BackgroundTasks,
    form_data: FormData = Depends(form_body_with_large_limits),
):
        file: Optional[UploadFile] = form_data.get("file")
        download_type_str: str = form_data.get("field_1", "default_field_value")
        ...
```
 It works

**Comment by JaeWangL** [comment]
final my usage is 
```
async def form_body_with_large_limits(request: Request) -> FormData:
    form = await request.form(max_part_size=50 * 1024 * 1024)
    return form
    
  @post()
 async def upload_zip(
   background_tasks: BackgroundTasks,
   form_data: FormData = Depends(form_body_with_large_limits),
):
       file: Optional[UploadFile] = form_data.get("file")
       download_type_str: str = form_data.get("field_1", "default_field_value")
       ...
```

**Comment by wszgrcy** [comment]
According to the source code, this should be a bug. Although parameters are set, the default parameters are always overwritten And there is no place to pass on references either
<img width="2427" height="480" alt="image" src="https://github.com/user-attachments/assets/e8aa9b5a-bc03-4a62-9571-fdc9879cc2a6" />
<img width="1784" height="1626" alt="image" src="https://github.com/user-attachments/assets/0cf63677-c252-4195-82e7-f464977bdf3b" />
<img width="2090" height="1434" alt="image" src="https://github.com/user-attachments/assets/b7b023e9-00e2-4380-a369-eb4f3d0a3424" />
<img width="2231" height="680" alt="image" src="https://github.com/user-attachments/assets/5b13ef58-5ab4-46a7-97aa-275c66e8566f" />
<img width="1522" height="398" alt="image" src="https://github.com/user-attachments/assets/c6463f47-94e2-4a2b-8f3c-5ae525b98b7d" />
